### PR TITLE
cmd/snap-update-ns: allow switching to user mount namespace

### PR DESCRIPTION
**Pull request by zyga**
When snap-update-ns is invoked with the (currently unused by snapd)
command line option `-u` it is an indication that we are going to run as
the specified user ID. In addition the option `--from-snap-confine` is
no longer attached to `--user-fstab` and now only implies that namespace
has already been set and that lock is already established.

This allows snap-update-ns to be used from two different contexts to
achieve distinct tasks:

    1) When invoked from snap-confine, it is always paired with
    `--from-snap-confine`. This, as mentioned, indicates that the mount
    namespace is already unshared and that locks are held by
    snap-confine.

    In this mode we can be called with no further options, to initialize
    the system-wide snap mount namespace or with `--user-mounts` to
    initialize the per-user snap mount namespace. Note that at this time
    `-u` is redundant because snap-confine already runs with the
    appropriate user ID.

    2) When invoked from snapd it is invoked with or without `-u` and
    `--user-mounts`. When those options are absent `snap-update-ns` will
    acquire appropriate locks, freeze running processes of the affected
    snap and perform an update to existing system-wide mount namespace
    that is joined, as appropriate. When those options are used the same
    flow will happen with the exception that per-user mount namespace
    will be joined and updated, running as the specified user ID.

This allows using snap-update-ns to update existing, persisted per-user
mount namespaces.

Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

File: cmd/snap-update-ns/bootstrap.c [hunk 1]
```diff
@@ -76,20 +83,10 @@ static int setns_into_snap(const char *snap_name)
...
+static int switch_to_specific_privileged_user(uid_t real_uid, gid_t real_gid)
 {
-	uid_t real_uid;
-	gid_t real_gid;
-
-	real_uid = getuid();
-	if (real_uid == 0) {
-		// We're running as root: no need to switch IDs
-		return 0;
-	}
```
Comment (jdstrand) [inline cmd/snap-update-ns/bootstrap.c]: Why is this not needed any more? (I'm sorry, I'm finding the argument processing a little hard to follow in terms of what is absolutely allowable and what is implied (process_arguments() from bootstrap.c would ideally document all the valid invocations and error on invalid combinations as well as when the snap name is required and when it isn't). I understand that in bootstrap this is only called when uid != 0, but why not leave the `if (real_uid == 0) { return 0; }` check for future-proofing?

File: cmd/snap-update-ns/bootstrap.c [hunk 2]
```diff
@@ -46,13 +46,20 @@ int bootstrap_errno = 0;
 const char *bootstrap_msg = NULL;
 
 // setns_into_snap switches mount namespace into that of a given snap.
-static int setns_into_snap(const char *snap_name)
+// If the uid argument is non-zero the mount namespace of a specific user is joined.
+// Otherwise the system-wide mount namespace of the provided snap is used.
```
Comment (jdstrand) [inline cmd/snap-update-ns/bootstrap.c]: I suggest:
```
// If the uid argument is non-zero the mount namespace of the provided snap
// for the specific user is joined. Otherwise the system-wide mount namespace
// of the provided snap is used.
```

File: cmd/snap-update-ns/bootstrap.c [hunk 3]
```diff
@@ -486,14 +483,15 @@ void bootstrap(int argc, char **argv, char **envp)
...
-	} else if (snap_name != NULL && should_setns) {
-		setns_into_snap(snap_name);
+	if (should_setns && snap_name != NULL) {
+		setns_into_snap(snap_name, process_user_fstab ? uid : 0);
 		// setns_into_snap sets bootstrap_{errno,msg}
 	}
+	if (process_user_fstab && uid != 0) {
+		switch_to_specific_privileged_user(uid, getgid());
+		// switch_to_privileged_user sets bootstrap_{errno,msg}
+	}
```
Comment (jdstrand) [inline cmd/snap-update-ns/bootstrap.c]: Maybe it's just me, but the combination of processing the args in a different file, this and the previous conditional and using the ternary operator here makes this all harder to understand than it could be. That said, the code apears to be correct. I don't have a better suggestion to make it easier to understand, which might be an indication of how snap-update-ns' invocations are too complex... (I can't help but think this could be made simpler... it's difficult to verify the behavior).

File: cmd/snap-update-ns/bootstrap.c [hunk 4]
```diff
@@ -486,14 +483,15 @@ void bootstrap(int argc, char **argv, char **envp)
...
-	unsigned long uid = 0;
+	unsigned long uid = getuid();
 	process_arguments(argc, argv, &snap_name, &should_setns,
 			  &process_user_fstab, &uid);
-	if (process_user_fstab) {
-		switch_to_privileged_user();
-		// switch_to_privileged_user sets bootstrap_{errno,msg}
-	} else if (snap_name != NULL && should_setns) {
-		setns_into_snap(snap_name);
+	if (should_setns && snap_name != NULL) {
```
Comment (jdstrand) [inline cmd/snap-update-ns/bootstrap.c]: Perhaps add a comment here:
```
// If needed, enter the snap's namespace using either the user fstab with our uid
// or the system wide snap fstab with uid 0
```

File: cmd/snap-update-ns/bootstrap.c [hunk 5]
```diff
@@ -486,14 +483,15 @@ void bootstrap(int argc, char **argv, char **envp)
...
+		setns_into_snap(snap_name, process_user_fstab ? uid : 0);
 		// setns_into_snap sets bootstrap_{errno,msg}
 	}
+	if (process_user_fstab && uid != 0) {
```
Comment (jdstrand) [inline cmd/snap-update-ns/bootstrap.c]: Perhaps add a comment here:
```
// If we are processing a user fstab, then drop privileges
```

File: cmd/snap-update-ns/bootstrap.c (hunk 4 above)
Comment (pedronis) [reply cmd/snap-update-ns/bootstrap.c]: A snap should always be set, no?  it would seem clearer if the snap_name != NULL would guard any action in the last part of bootstrap

File: cmd/snap-update-ns/bootstrap.c (hunk 3 above)
Comment (pedronis) [reply cmd/snap-update-ns/bootstrap.c]: Btw the name switch_to_privileged_user is a bit strange, the user is not privileged, is that we switch to it but keep/reacquire some our caps afaiu

Comment (zyga) [reply cmd/snap-update-ns/bootstrap.c]: I renamed the function now.

Comment (zyga) [reply cmd/snap-update-ns/bootstrap.c]: @jdstrand I have a plan to simplify this but I want to complete the several ongoing changes that affect this area first. The plan will involve a new binary, `snap-boostrap-ns` that will both take over responsibility of parts of `snap-confine` and also simplify command line handling of `snap-update-ns`, which will now only be invoked by `snapd` for actual updates.

**Comment by zyga** [comment]
I will return to this branch once some of the related branches land. As with some of the other open PS, it is a result of splitting a complex PR and cannot be easily tested in isolation. As such I will resume this work only after the main refactoring branch is merged.

**Comment by zyga** [comment]
This is getting close now, most of the prerequisites have landed and I will be able to re-assemble the user-mount patch in its entirety soon.

**Comment by zyga** [comment]
Closing to deal with master woes and refocus. Please make sure not to remove this branch though :)
